onnxruntime/core/graph/graph_utils.cc

```diff
@@ -191,51 +193,121 @@ static void UpdateImplicitInputNameInSubgraph(Node& node,
   }
 }
 
-bool RemoveSingleInSingleOutNode(Graph& graph, Node& node) {
-  if (!IsSingleInSingleOutNode(node)) {
-    return false;
+/** Takes as input an outgoing edge and the name of an input of a node of the given graph.
+    Checks if the outgoing edge is used to provide input to a subgraph. If so, it updates the input of the subgraph
+    to use the arg_name of the incoming edge. This process is needed before removing the node from the graph.
+    Returns false if the input update cannot be performed safely. */
+static bool CheckAndUpdateImplicitInputInSubgraph(Graph& graph,
+                                                  const std::string& input_name,
+                                                  const GraphEdge& output_edge) {
+  if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+    Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+    if (CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_name)) {
+      UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_name);
+    } else {
+      // we can't safely remove this node
+      return false;
+    }
+  }
+  return true;
+}
+
+/** Removes output edges of a given node and returns a vector of these edges. */
+static std::vector<GraphEdge> RemoveAndReturnOutputEdges(Graph& graph, Node& node) {
```

Comment:
>RemoveAndReturnOutputEdges [](start = 30, length = 26)

Would be more flexible if this is two separate methods (one to get the output edges, one to remove a set of edges).

Reply:
I created a method called GetNodeOutputEdges, and reuse it in both RemoveNodeOutputEdges and RemoveAndReturnOutputEdges. I purposely did not make it public, because I dont want to expose the GraphEdge struct. It is very useful for the graph_utils, but I am afraid it will cause confusioin if we have methods returning it, as we already have the EdgeEnd in the Graph.
Let me know what you think.

Reply:
Doesn't need to be public. But why couple removal and fetching output edges in one function? Could there instead by a more generic RemoveEdges that takes a collection of GraphEdge, and in this case the user calls GetNodeOutputEdges to get that collection and passes that into RemoveEdges?

Reply:
My idea was that you would never have a set of GraphEdges to call RemoveEdges. Only GetNodeOutputEdges will be giving it to you, so I didn't even want to allow someone to call RemoveEdges with a set of GraphdEdge (even if it is an internal method).
But it is a small detail, I can do it as you suggest.